src/Tasks/FileIO/GetFileHash.cs

```diff
@@ -0,0 +1,126 @@
+
+        /// <summary>
+        /// The metadata name where the base64 encoded hash is store in each item. File hash is in hex. Defaults to "FileHashBase64".
+        /// </summary>
+        public string MetadataNameBase64 { get; set; } = "FileHashBase64";
+
+        /// <summary>
+        /// The hash of the file in hex. This is only set if there was one item group passed in.
+        /// </summary>
+        [Output]
+        public string Hash { get; set; }
+
+        /// <summary>
+        /// The hash of the file base64 encoded. This is only set if there was one item group passed in.
+        /// </summary>
+        [Output]
+        public string HashBase64 { get; set; }
+
+        /// <summary>
+        /// The input files with additional metadata set to include the file hash.
+        /// </summary>
+        [Output]
+        public ITaskItem[] Items { get; set; }
+
+        public override bool Execute()
+        {
+            if (!SupportsAlgorithm(Algorithm))
+            {
+                Log.LogErrorFromResources("FileHash.UnrecognizedHashAlgorithm", Algorithm);
+                return false;
+            }
+
+            foreach (var file in Files)
+            {
+                if (!File.Exists(file.ItemSpec))
+                {
+                    Log.LogErrorFromResources("FileHash.FileNotFound", file.ItemSpec);
+                    continue;
+                }
+
+                var hash = ComputeHash(Algorithm, file.ItemSpec);
+                file.SetMetadata("FileHashAlgoritm", Algorithm);
+                file.SetMetadata(MetadataName, ConversionUtilities.ConvertByteArrayToHex(hash));
+                file.SetMetadata(MetadataNameBase64, Convert.ToBase64String(hash));
+            }
+
+            Items = Files;
+
+            if (Files.Length == 1)
+            {
+                Hash = Files[0].GetMetadata(MetadataName);
+                HashBase64 = Files[0].GetMetadata(MetadataNameBase64);
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        internal static bool SupportsAlgorithm(string algorithmName) => _supportedAlgorithms.Contains(algorithmName);
+
+        internal static byte[] ComputeHash(string algorithmName, string filePath)
+        {
+            using (var stream = File.OpenRead(filePath))
+            using (var algorithm = CreateAlgorithm(algorithmName))
+            {
+                return algorithm.ComputeHash(stream);
+            }
+        }
+
+        private static readonly HashSet<string> _supportedAlgorithms
+            = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+            {
+                "SHA256",
+                "SHA384",
+                "SHA512",
+            };
+
+        private static HashAlgorithm CreateAlgorithm(string algorithmName)
+        {
+            switch (algorithmName.ToUpperInvariant())
+            {
+                case "SHA256":
+                    return SHA256.Create();
```

Comment:
Can these objects be reused for the entire execution or do they carry state and need to be created for each ComputeHash call?

Reply:
Aren't task instances recreated?

Reply:
Yes, but hash algorithm objects are created for each file to hash. Probably not a problem unless thousands of files get hashed.

Reply:
IIUC HashAlgorithm objects carry state so need to be recreated for each file they are hashing.